### PR TITLE
Add missing signals for SpinButton

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
cc @EPashkin

### CODE REVIEW

File: src/signal.rs [hunk 1]
```diff
@@ -1706,34 +1707,111 @@ mod editable {
...
+                let f: Box<Box<Fn(&SpinButton, ScrollType) + 'static>> =
+                    Box::new(Box::new(change_value_func));
+                connect(self.to_glib_none().0, "change_value",
+                        transmute(change_trampoline as usize), Box::into_raw(f) as *mut _)
+            }
+        }
 
-    impl SpinButton {
-        pub fn connect_input<F: Fn(&SpinButton, &mut f64) -> Result<bool, ()> + 'static>(&self, f: F) -> u64 {
+        fn connect_input<F>(&self, f: F) -> u64
```
Comment (GuillaumeGomez) [inline src/signal.rs]: Could you take a look here @gkoz?

File: src/signal.rs [hunk 2]
```diff
@@ -1706,34 +1707,111 @@ mod editable {
...
+
+    unsafe extern "C" fn change_trampoline(this: *mut GtkSpinButton,
+                                           scroll: ScrollType,
+                                           f: &Box<Fn(&SpinButton, ScrollType) + 'static>) {
+        callback_guard!();
+        f(&from_glib_none(this), scroll)
     }
 
-    unsafe extern "C" fn input_trampoline(this: *mut GtkSpinButton, new_value: *mut c_double, f: gpointer) -> c_int {
+    unsafe extern "C" fn input_trampoline(this: *mut GtkSpinButton,
```
Comment (GuillaumeGomez) [inline src/signal.rs]: And here too please @gkoz.

File: src/signal.rs [hunk 3]
```diff
@@ -1706,34 +1707,111 @@ mod editable {
     }
 }
 
+pub trait SpinButtonSignals {
```
Comment (EPashkin) [inline src/signal.rs]: IMHO trait don't needed

Comment (GuillaumeGomez) [inline src/signal.rs]: It's just to be consistent with the other signals in this file.

Comment (EPashkin) [inline src/signal.rs]: Them may be `mod gl_area` need fixing?

Comment (GuillaumeGomez) [inline src/signal.rs]: If it's in this file, then yes. Want me to do it in this PR as well? Or I open a new one?

Comment (EPashkin) [inline src/signal.rs]: It in this file, only one without `xxSignals`.
Not sure, that it really need be fixed. IMHO better wait for gkoz.

**Comment by gkoz** [comment]
Regarding the `input` signal. Should it be `Fn(&SpinButton) -> Option<Result<f64, ()>>`?
That is, first of all the signal handler has a choice of either trying to convert the value or just reacting to the signal. If it doesn't convert it `None` is returned (this corresponds to `Inhibit(false)`). If it does attempt conversion, then it's `Some(result)` (inhibiting the default handler).

**Comment by EPashkin** [comment]
@gkoz Your closure variant much better.
@GuillaumeGomez Change it please, and I fix example.

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/signal.rs [hunk 4]
```diff
@@ -1706,34 +1707,117 @@ mod editable {
...
-            Err(_) => GTK_INPUT_ERROR,
+        f(&from_glib_none(this), scroll)
+    }
+
+    unsafe extern "C" fn input_trampoline(this: *mut GtkSpinButton,
+                                          new_value: *mut c_double,
+                                          f: &Box_<Fn(&SpinButton) -> Option<Result<f64, ()>> + 'static>)
+                                          -> c_int {
+        callback_guard!();
+        if let Some(v) = f(&from_glib_none(this)) {
```
Comment (EPashkin) [inline src/signal.rs]: Not better add one  `match` ?

**Comment by EPashkin** [comment]
Also needed ignoring generating signals and Regen.

File: src/signal.rs (hunk 4 above)
Comment (GuillaumeGomez) [inline src/signal.rs]: You gave me an idea actually! Hold on!

**Comment by GuillaumeGomez** [comment]
@EPashkin: In just one match now. :D

File: src/signal.rs [hunk 5]
```diff
@@ -1706,34 +1707,114 @@ mod editable {
...
             unsafe {
-                let f: Box_<Box_<Fn(&SpinButton, &mut f64) -> Result<bool, ()> + 'static>> = Box_::new(Box_::new(f));
+                let f: Box_<Box_<Fn(&SpinButton) -> Option<Result<f64, ()>> + 'static>> = Box_::new(Box_::new(f));
                 connect(self.to_glib_none().0, "input",
                         transmute(input_trampoline as usize), Box_::into_raw(f) as *mut _)
             }
         }
+
+        fn connect_output<F>(&self, output_func: F) -> u64
+        where F: Fn(&SpinButton) -> bool + 'static {
```
Comment (gkoz) [inline src/signal.rs]: It looks like the return type here has to be `Inhibit`.

**Comment by EPashkin** [comment]
@GuillaumeGomez It much better.

File: src/signal.rs (hunk 5 above)
Comment (GuillaumeGomez) [inline src/signal.rs]: Then `Inhibit` it'll be.

Comment (EPashkin) [inline src/signal.rs]: Why `Inhibit` ?
```
Returns
TRUE if the value has been displayed
```

Comment (EPashkin) [inline src/signal.rs]: Due `_gtk_boolean_handled_accumulator` in code?

Comment (GuillaumeGomez) [inline src/signal.rs]: So... Do I change it or not?

Comment (gkoz) [inline src/signal.rs]: > Due `_gtk_boolean_handled_accumulator` in code?

Yes, the documentation fails to make the semantics clear sometimes but the use of this accumulator strongly suggests `Inhibit`.

**Comment by GuillaumeGomez** [comment]
I regenerated and added the `Inhibit` return.

**Comment by gkoz** [comment]
Out of curiosity why does the generator not generate `connect_wrapped`? The signature seems basic and I don't see any `ignore` directives.

**Comment by gkoz** [comment]
Um looks like it actually does generate four signals.

**Comment by EPashkin** [comment]
@GuillaumeGomez fix this please, it must be `[[object.signal]]`
```
[[object]]
name = "Gtk.StatusIcon"
status = "generate"
    [[object.function]]
    pattern = ".+" #not all generated
    ignore = true
```

**Comment by EPashkin** [comment]
"Regen" again with different version of `gir` :(

**Comment by GuillaumeGomez** [comment]
Updated `Gir.toml` file and regenerated.

**Comment by EPashkin** [comment]
Thanks, but add "Update gir" commit please if you regen with `gir (af5277e)`

**Comment by GuillaumeGomez** [comment]
I don't have the "original' regen commit, the new one is with `gir (af5277e)`.

**Comment by EPashkin** [comment]
Seems we have some misunderstanding.
I asked you updating git sub-module, same as you do in 64fcc0730c8a84b239ad3ea9c210aebd62e23c35

**Comment by GuillaumeGomez** [comment]
Totally misunderstood, indeed. ^^'

**Comment by EPashkin** [comment]
Other misunderstanding: we need ignore generation of spinbutton's signals.

**Comment by GuillaumeGomez** [comment]
All updated.

**Comment by EPashkin** [comment]
Thanks. @gkoz merge this please

**Comment by gkoz** [comment]
Thanks!

**Pull request merged**